join.spec.md

```diff
@@ -100,204 +101,347 @@ flowchart TB
+
+```graphql example -- Provided fields
+# Supergraph schema
+type Query {
+  fieldA: X @join__field(graph: A, provides: "usuallyBField")
+  fieldB: X @join__field(graph: B)
+}
+
+type X @join__owner(graph: B) {
+  usuallyBField: String @join__field(graph: B)
+}
+
+# Operation
+{ fieldB { usuallyBField } }
+# Generated subgraph operations
+## On B
+{ fieldB { usuallyBField } }
+
+# Operation
+{ fieldA { usuallyBField } }
+# Generated subgraph operations
+## On A
+{ fieldA { usuallyBField } }
 ```
 
-:::[example](photos.graphql#join__Graph) -- Using {@join__graph} to declare subgraph metadata on the {join__Graph} enum values.
+## Fields on value types
 
-##! @join__type
+Some types have the property that all of their fields can be resolved by *any* subgraph that can resolve a field returning that type. These types are called *value types*. (Imagine a type `type T { x: Int, y: String }` where every resolver for a field of type `T` actually produces an object like `{x: 1, y: "z"}`, and the resolvers for the two fields on `T` just unpack the values already in the object.) In a supergraph schema, a type is a value type if it does not have a [{@join__owner}](#@join__owner) directive on it.
 
-Join a type to a subgraph, optionally providing an entity key.
+```graphql example -- Value types
+# Supergraph schema
+type Query {
+  fieldA: X @join__field(graph: A)
+  fieldB: X @join__field(graph: B)
+}
 
-```graphql definition
-directive @join__type(
-  graph: join__Graph!
-  key: String
-) repeatable on OBJECT | INTERFACE
+type X {
+  anywhere: String
+}
+
+# Operation
+{ fieldA { anywhere } }
+# Generated subgraph operations
+## On A
+{ fieldA { anywhere } }
+
+# Operation
+{ fieldB { anywhere } }
+# Generated subgraph operations
+## On B
+{ fieldB { anywhere } }
 ```
 
-// TODO: rename portal query?
-Keys will be passed as `representations` within a [portal query](#portal-query) to [port](#portability) a selection set between subgraphs.
+## Owned fields on owned types
 
-:::[example](photos.graphql#Image) -- Using {@join__type} to specify subgraph keys
+We've finally reached the most interesting case: a field that must be resolved by an operation on a different subgraph from the subgraph on which its parent field was resolved. In order to do this, we need a way to tell the subgraph to resolve that parent object. We do this by defining a special root field in the subgraph's schema: `Query._entities(representations: [_Any!]!): [_Entity]!`. This field takes a list of "representations", which are JSON objects containing a `__typename` field and some other fields, and returns a list of the same length of the corresponding objects resulting from looking up the representations in an application-dependent way.
```

Comment:
This is a lot to drop on someone. What's an `_Any` and `_Entity`? Should we clarify that "some other fields" are the keys required to resolve the entity?

I'm also not sure if it belongs in the spec. I think there is a disclaimer about how the examples provided are not necessarily prescriptive, but this is an implementation detail of specifically how Apollo Federation works. This is getting into the details of query planning / execution which I'm not sure this spec is meant to be responsible for.

This is totally up for discussion but wanted to express my initial reaction. I understand that our implementation is useful in demonstrating / explaining the concepts. If we decide we want to do that, a note beforehand would substantially quiet my concerns.

Reply:
So I get the part that's "it doesn't have to be Apollo Federation" but I interpreted that to mean "you could have just written this supergraph schema by hand, there's no promise that there were subgraph schemas with `@key` on it".

I don't see how this spec describes something that could be used to drive a router other than by using `Query._entities`? I mean what else are you supposed to do with `key` and `requires`? Certainly I can imagine a future where the query isn't hardcoded and there are other ways to get non-root fields from subgraphs, though that would have to be an extension of the format?  Like, it does feel to me that `Query._entities` is a hard-coded part of the "router algorithm" for this version of the join spec (note that it was in the part that I deleted too).

I gave restructuring this a bit of a shot in d6c90c0. Not sure if it helps